packages/apps/dashboard/server/src/modules/details/details.service.ts

```diff
@@ -148,40 +153,88 @@ export class DetailsService {
     return result;
   }
 
-  public async getLeadersByChainId(
+  public async getLeaders(
     chainId?: ChainId,
-    take?: number,
+    orderBy?: LeadersOrderBy,
+    orderDirection?: OrderDirection,
+    first?: number,
+    skip?: number,
   ): Promise<LeaderDto[]> {
-    const chainIds = !chainId
-      ? await this.networkConfig.getAvailableNetworks()
-      : [chainId];
+    const chainIds = chainId
+      ? [chainId]
+      : await this.networkConfig.getAvailableNetworks();
 
-    let allLeadersData: ILeader[] = [];
-    for (const id of chainIds) {
-      const leadersData = await OperatorUtils.getLeaders({ chainId: id });
-      allLeadersData = allLeadersData.concat(
-        leadersData.filter((leader) => leader.amountStaked > 0 && leader.role),
+    const results = await Promise.all(
+      chainIds.map(async (id) => {
+        const filter = this.createLeadersFilter(
+          id,
+          orderBy,
+          orderDirection,
+          first,
+          skip,
+        );
+        const [leaders, reputations] = await Promise.all([
+          OperatorUtils.getLeaders(filter),
+          this.fetchReputations(id),
+        ]);
+        return { leaders, reputations };
+      }),
+    );
+
+    const leadersData = results.flatMap(({ leaders }) => leaders);
+    const reputations = results.flatMap(({ reputations }) => reputations);
+
+    const leaders = leadersData
+      .filter((leader) => leader.role)
+      .map((leader) =>
+        plainToInstance(LeaderDto, leader, { excludeExtraneousValues: true }),
       );
-    }
 
-    allLeadersData.sort((a, b) =>
-      BigInt(a.amountStaked) >= BigInt(b.amountStaked) ? -1 : 1,
-    );
-    if (take && take > 0) {
-      allLeadersData = allLeadersData.slice(0, take);
+    this.assignReputationsToLeaders(leaders, reputations);
+
+    if (orderBy === LeadersOrderBy.REPUTATION) {
+      return leaders
+        .sort((a, b) => {
+          const rankA = REPUTATION_RANK[a.reputation] ?? -1;
+          const rankB = REPUTATION_RANK[b.reputation] ?? -1;
+          return orderDirection === OrderDirection.DESC
+            ? rankA - rankB
+            : rankB - rankA;
+        })
+        .slice(skip ?? 0, (skip ?? 0) + (first ?? leaders.length));
     }
-    const leaders = allLeadersData.map((leader) =>
-      plainToInstance(LeaderDto, leader, {
-        excludeExtraneousValues: true,
-      }),
-    );
+    return leaders;
```

Comment:
If `orderBy` is one of the leader fields (i.e. not "reputation"), then when querying leaders for all chains we need to:
1. Get N (as requested page size) from each chain
2. Merge all results to a single array
3. Sort this array per order criteria
4. Slice it to N results

This way you will address they issue of the order when retrieving data from multiple chains.

As an alternative, I can suggest to remove the option to get leaders across all chains and make `chainId` required parameter (that shouldn't be `ALL`/`-1`). I don't see a scenario for the user when they might want to see "leaders" across all chains

wdyt, @portuu3 ?

Reply:
lgtm